adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/sasl/AmqpAdapterSaslAuthenticatorFactory.java

```diff
@@ -0,0 +1,214 @@
+            } else if (!AuthenticationConstants.MECHANISM_PLAIN.equals(remoteMechanisms[0])) {
+                LOG.debug("SASL mechanism [{}] is currently not supported by the AMQP adapter", remoteMechanisms[0]);
+                sasl.done(SaslOutcome.PN_SASL_AUTH);
+                completionHandler.handle(Boolean.TRUE);
+            } else {
+                final String remoteMechanism = remoteMechanisms[0];
+                LOG.debug("client device wants to authenticate using SASL [mechanism: {}, host: {}, state: {}]",
+                        remoteMechanism, sasl.getHostname(), sasl.getState());
+
+                final Future<Device> deviceAuthTracker = Future.future();
+                deviceAuthTracker.setHandler(outcome -> {
+                    if (outcome.succeeded()) {
+
+                        final Device authenticatedDevice = outcome.result();
+                        protonConnection.attachments().set(Constants.KEY_CLIENT_DEVICE, Device.class,
+                                authenticatedDevice);
+                        succeeded = true;
+                        sasl.done(SaslOutcome.PN_SASL_OK);
+
+                    } else {
+                        LOG.debug("validation of credentials failed: " + outcome.cause().getMessage());
+                        sasl.done(SaslOutcome.PN_SASL_AUTH);
+
+                    }
+                    completionHandler.handle(Boolean.TRUE);
+                });
+
+                final byte[] saslResponse = new byte[sasl.pending()];
+                sasl.recv(saslResponse, 0, saslResponse.length);
+
+                verifyPlain(remoteMechanism, saslResponse, deviceAuthTracker.completer());
+            }
+        }
+
+        @Override
+        public boolean succeeded() {
+            return succeeded;
+        }
+
+        private void verifyPlain(final String remoteMechanism, final byte[] saslResponse,
+                final Handler<AsyncResult<Device>> completer) {
+
+            getFields(saslResponse).compose(fields -> {
+                final DeviceCredentials credentials = UsernamePasswordCredentials
+                        .create(fields[0], fields[1], config.isSingleTenant());
+                if (credentials == null) {
+                    // adapter is configured for multi-tenancy but credentials
+                    // does not comply with the pattern [<authId>@<tenantId>]
+                    completer.handle(Future.failedFuture(new CredentialException(
+                            "username does not comply with expected pattern [<authId>@<tenantId>]")));
+                } else {
+                    authProvider.authenticate(credentials, completer);
+                }
+                return Future.future();
+            }).recover(t -> {
+                completer.handle(Future.failedFuture(t));
+                return Future.failedFuture(t);
+            });
+
+        }
+
+        private Future<String[]> getFields(final byte[] saslResponse) {
```

Comment:
IMHO this method should rather return a `Future<UsernamePasswordCredentials>` ...